### ISSUE TITLE
Scatter-add in bfloat16 is much slower than in float32 when contention is high

### ISSUE THREAD

**Issue by jaro-sevcik** [openxla/xla#26371]
Reproducer is below, it illustrates ~20x slowdown when using bfloat16. This was with JAX 0.6.0 on GH200). The repro was simplified from an example where the scatter-add comes from a backward pass of an embedding layer with a small embedding vector, so this pattern is not completely made up.

```python
import jax
import jax.numpy as jnp
from time import time

UPDATES=1000000
SIZE=10
BATCH=10

key = jax.random.PRNGKey(1)
selectors = jax.random.randint(key, (UPDATES,), 0, SIZE, dtype=jnp.int32)

def measure(dt):
    print("==== Dtype: ", dt)
    acc = jnp.zeros((BATCH, SIZE), dtype=dt)
    updates = jnp.ones((BATCH, UPDATES), dtype=dt)
    
    @jax.jit
    def scatter_add(acc, selectors, updates):
        return acc.at[:, selectors].add(updates)
    
    for i in range(2):
        start = time()
        scatter_add(acc, selectors, updates).block_until_ready()
        dt = (time() - start) * 1000.0
        print(f'Scatter time (iteration {i}): {dt:.3f}ms')

measure(jnp.float32)
measure(jnp.bfloat16)
```

Result:
```
==== Dtype:  <class 'jax.numpy.float32'>
Scatter time (iteration 0): 48.330ms
Scatter time (iteration 1): 0.960ms
==== Dtype:  <class 'jax.numpy.bfloat16'>
Scatter time (iteration 0): 73.820ms
Scatter time (iteration 1): 18.155ms
```

**Comment by jaro-sevcik** [comment]
When looking at the profile, it looks like the main problem is that float32 atomic-add is much faster than bf16 atomic-add, especially when there is a lot of collisions (such as in the example above where we are adding a million numbers to just ten accumulators).

For the high collision case, it would be faster to reduce rather than atomic-add:
```python
import jax
import jax.numpy as jnp
from time import time

UPDATES=1000000
SIZE=10
BATCH=10

key = jax.random.PRNGKey(1)
selectors = jax.random.randint(key, (UPDATES,), 0, SIZE, dtype=jnp.int32)

def measure(dt):
    print("==== Dtype: ", dt)
    acc = jnp.zeros((BATCH, SIZE), dtype=dt)
    updates = jnp.ones((BATCH, UPDATES), dtype=dt)
    
    @jax.jit
    def scatter_add(acc, selectors, updates):
        return acc.at[:, selectors].add(updates)

    @jax.jit
    def reduce_add(acc, selectors, updates):
        mask = jnp.arange(SIZE).reshape((1, SIZE, 1)) == selectors.reshape((1, 1, UPDATES))
        result = jnp.sum(mask * jnp.expand_dims(updates, -2), axis=-1)
        return result
    
    for i in range(2):
        start = time()
        scatter_add(acc, selectors, updates).block_until_ready()
        dt = (time() - start) * 1000.0
        if i > 0:
            print(f'Scatter time (iteration {i}): {dt:.3f}ms')

        start = time()
        reduce_add(acc, selectors, updates).block_until_ready()
        dt = (time() - start) * 1000.0
        if i > 0:
            print(f'Reduce time (iteration {i}): {dt:.3f}ms')

measure(jnp.float32)
measure(jnp.bfloat16)
```

Output:
```
==== Dtype:  <class 'jax.numpy.float32'>
Scatter time (iteration 1): 0.962ms
Reduce time (iteration 1): 0.193ms
==== Dtype:  <class 'jax.numpy.bfloat16'>
Scatter time (iteration 1): 17.996ms
Reduce time (iteration 1): 0.153ms
```

However, the advantage of reduce quickly diminishes with fewer collisions. E.g., for `SIZE=1000`, we get:

```
==== Dtype:  <class 'jax.numpy.float32'>
Scatter time (iteration 1): 0.217ms
Reduce time (iteration 1): 6.318ms
==== Dtype:  <class 'jax.numpy.bfloat16'>
Scatter time (iteration 1): 1.780ms
Reduce time (iteration 1): 5.424ms
```

**Comment by penumallas1261** [comment]
Internal reproducer submitted in cl/773527230.
